GUI upgrade + general code improvements + new farms: FoW Toc, Gemstone Margonite/Torment, Underworld, Glint, Asura, Minotaurs

Gahais opened the pull request:
hi :)

Here is second pull request. You don't have analyze it now if you don't have free time. 
New pull request is in new branch in your repository - GUI-Gemstone-Margonite-FoW-ToC-Improvements
Thank You again for creating this repository so that all Guild Wars bots can be professionally managed on GitHub instead of dealing with problematic zip files.
I see that you have made some interesting fixes and improvements recently, I merged them into this pull request :)
If you have some uncommitted changes then we can try to merge them somehow into this pull request


Here are my humble suggestions, you can tell me if it is fine, and I can refactor that if necessary
- Changed HP into HealthPercent and MaxHP into MaxHealth for clarity and to make it consistent with energy, EnergyPercent and MaxEnergy in $agentStructTemplate. Values are generally the same HealthPercent and EnergyPercent are in range [0..1] and Maxhealth is like 480 and MaxEnergy like 50, so it can be clear this way
- Changed Id1, Id2, etc., into SkillId1, SkillId2, etc. in $skillbarStructTemplate to make it more clear what this Id stands for, SkillId is also used in other places of the code I think
- renamed Stopped GUI state into Uninitialized at start of the app so that it changes into Initialized after Authentification(), we can change naming back, if you don't like it
- renamed Gemstone farm into Gemstones to make it different from single gemstone farms
- renamed all custom colors in GUI with inbuilt color names from AutoIT, for more code clarity, I think colors might be quite similar to original ones, we can revert that if you don't like it
- renamed HM to Hard Mode in GUI for more clarity, I think
- changed 1 to True, 0 to False for Opt() calls, Imho this can be more clear
- changed 1, 3, 6 into `$ID_Allegiance_Team, $ID_Allegiance_Foe, $ID_Allegiance_Npc` + added unused constants `$ID_Allegiance_Animal = 2, $$ID_Allegiance_Spirit = 4, $$ID_Allegiance_Minion = 5` to have less magical constants
- restored these party check functions like IsPartyWiped() for convenience. But we can remove them back if you don't like them
- reduced sleep timeout in LoadSkillTemplate() to make faster build loading especially for these automatic team setups. I hope that this isn't a bad idea, what do you think? So far it worked fine for me
- maybe also use $STUCK = 3 return result from functions for example when player is stuck by mobs during running or when max time for bot has elapsed

Added GUI improvements. Couldn't split big GUI commit into small ones. It's commit acf1e85
- increased GUI size to be able to fir more item counters and automatic team setup in Run Options and also swapped place of some elements, what do you think about it? We can refactor that. But, I think when merging JackLinesMatthews fork the GUI would need to be refactored again, either way, I think. Below should be a screenshot
- added success ratio to quickly see the percentage of success,fails
- Added missing item counters like Jade bracelets for Jade Brotherhood farm, Obsidian Shards for new FoW ToC farm, lockpicks which can drop sometimes in some farms like in Vaettirs farm
- fix for comboboxes. Disabling them during run to avoid being accidentally focused, scrolled and changed. Reenabling them after run finishes. Changed comboboxes so that can't write any text into them (except only for character combobox). Added limited options for bags combobox to 1,2,3,4,5. Fix of weird AutoIT problem that character and farm combobox don't update after first GUICtrlSetData, I'm not sure if this happens only for me
- Added automatic player and team setup that can be turned on and off with checkbox. Comboboxes for heroes and possibility to write build templates in hero textfields and save to configuration file. In team farms like vanquishes or FoW or SoO this can automatically load builds for team if checkbox is checked
- Added checkbox to choose if faction points should be donated or spent for purchase of amber chunks or jadeite shards for kurzick and luxon farms
- Added checkbox to use scrolls or gold or nothing in Pantheon bonus weeks for Fow and FoW ToC farms
- Added colorful Render button to be able to turn rendering on and off any time. But honestly when I tested this rendering turned off I noticed some weird problems and bot was probably more prone to crashes. Did you notice some weird problems when rendering is disabled?
For example in Vaettir farm when I disable rendering then player can correctly travel to Jaga Moraine but then can stop forever near the norn shrine


Added some fixes for effects, recharge of skills, skill usage and treating empty quest as Null
Added DoForArrayRows() function to make dealing with arrays more compact
Added universal function MoveAvoidingBodyBlock() that can be controlled with $options parameters similarly like MoveAggro() functions, because I noticed that quite similar movement functions were used in Raptors, Vaettirs, FoW ToC and single gemstones farms

Added new farms
- Fissure of Woe Tower of Courage farm, quite well working and fully tested
- Gemstone Margonite farm with possibility to use 4 professions, didn't test it for ranger though, I don't have ranger in Gate of Anguish :(
- Vaettir farm refactor with possibility to use 4 professions, fully tested

I noticed that I broke this Corsairs farm previously because its mission had identical map ID as outpost so updated it now similarly as gemstones farm that checks initial spawn coordinates of player in mission and resigns after finishing farm


Maybe we should replace all these calls like DllStructGetData() with GetStructData()? What do you think about it? This might increase code readability and the sllight cost of performance probably. I can replace all such calls in third pull request. Or maybe this is a bad idea

I see JackLinesMatthews has some interesting code in his fork. It's cool you integrated Boreal farm into BotShub. Have you been able to contact JackLinesMatthews? Is he planning to do a pull request of his code into BotsHub?
I can do a third pull request with integration of best ideas of JackLinesMatthews from his fork. I can write custom, compact commits in a new branch based on his code with yours and JackLinesMatthews permissions. I may have some ideas how to quickly integrate that. I can credit JackLinesMatthews for his ideas in the code and credtis.md accordingly. What do you think about that? I can ask @JackLinesMatthews

Below 2 screens of changed GUI
<img width="800" height="428" alt="screen1" src="https://github.com/user-attachments/assets/51746800-4fbf-4e91-a342-c362ca52a7a9" />
<img width="800" height="430" alt="screen2" src="https://github.com/user-attachments/assets/ebfc9f7d-2f2a-4629-8ef5-ed3aa94fb652" />

caustic-kronos commented on lib/GWA2.au3:
```diff
@@ -3195,6 +3195,7 @@ EndFunc
 
 
 ;~ Returns the target of an agent.
+;~ Caution, this function might not be working correctly
```
If you're sure of yourself concerning those functions not working, let's mark them with a tag (FIXME) so we find them easily

Gahais replied:
yeah, I tested these functions that they always return 0 or something like this, despite situation
ok, marked these 3 functions with FIXME:

JackLinesMatthews commented:
My discord username is jacko8167

caustic-kronos commented on src/Farm-Vaettirs.au3:
```diff
@@ -206,17 +231,7 @@ Func AggroAllMobs()
-	MoveAggroing(11670, -15457, 150)
-	MoveAggroing(12604, -15320, 150)
-	MoveAggroing(12476, -16157)
+	If DoForArrayRows($vaettirs, 15, 25, MoveAggroing) == $FAIL Then Return $FAIL
```
I'm not convinced by DoForArrayRows.
I get it, we set a practical function for repeated function use.
But we add an additional layer of abstraction and its only purpose (ultimately) is to do a for loop on the array we provide.
I think I'd rather keep a for loop for those places.
```
For $i = 15 To 25
	If MoveAggroing($vaettirs[$i][0], $vaettirs[$i][1]) == $FAIL Then Return $FAIL
Next
```

Gahais replied:
okay, you are the maintainer, you can decide
So I replaced this with for loops in commit 0d34c29d

caustic-kronos commented on src/Farm-Raptors.au3:
```diff
@@ -273,11 +279,10 @@ Func GetRaptors()
-	Local Static $unaggroedState = DllStructGetData($boss, 'TypeMap')
-	Return DllStructGetData($boss, 'TypeMap') <> $unaggroedState
+	Local $boss = GetBossFoe()
+	Return BitAND(DllStructGetData($boss, 'TypeMap'), 0x1) == 1
```
Is it only the first bit of TypeMap that corresponds to "aggroed" ?
If so good correction ! I remember when I rewrote raptors a long time ago the value could be different between runs but that was the whole value.

Gahais replied:
yes, there were different whole values but these seem to be bitmasks with various bits. Below should be some notes I made on different TypeMap values. I noticed that player or other humans in game can be in both stance. If they are out of combat like in city then they stand upright (TypeMap 0x0). But when they are in combat then they get bent forward (TypeMap 0x1) even when foe is far but within earshot range
But for other npcs like raptors there may not be 2 stances animations so raptor or other mobs can change from 0x0 to 0x1 without noticeable animations

2^0 = 0x0001 - attacking or attack stance of an agent. Bit on 1th position
2^13 = 0x2000 - usage of skill or manual attack by agent. Bit on 14th position
2^3 = = 0x8 - death of party member or foe. Bit on 4th position

TypeMap = 0x2 = 2 = corresponds to npcs in cities
TypeMap = 0x0 = 0 = corresponds to enemies, who don't do anything, for example when far away
TypeMap = 0x1 = 1 = corresponds to enemies, when they are in attack stance
TypeMap = 0x8 = 8 = corresponds to enemies in dead state
TypeMap = 0x2001 = 8193 = corresponds to enemies, during fights, when using skills/attacking

TypeMap = 0x20000 = 131072 = corresponds to party members. Party members that are idle and not using any skills
TypeMap = 0x20001 = 131073 = corresponds to attacking party members
TypeMap = 0x22001 = 139265 = corresponds to attacking party members and using skills

TypeMap = 0x421004 = 4329476 = corresponds to idle player, when not casting and not fighting
TypeMap = 0x421005 = 4329477 = corresponds to attacking player 
TypeMap = 0x423005 = 4337669 = corresponds to attacking player who is using a skill 
TypeMap = 0x42100C = 4329484 = corresponds to dead player 

TypeMap = 0x40000 = 262144 = corresponds to spirits created by ranger
TypeMap = 0x40001 = 262145 = corresponds to spirits created by ritualist. And also bone minions. Agents that can attack

caustic-kronos commented on conf/characters/Default Configuration.json:
```diff
@@ -6,23 +6,42 @@
+		"hero_6": "Norgu",
+		"hero_6_build": "OQNEAqwD2ycCwpmupXOIDcBQj",
+		"hero_7": "Razah",
+		"hero_7_build": "OQNEAqwD2ycCaCmupXOIDMEQj"
```
Also maybe we can rename tabs to take less room and have more of them:
main
run
loot
infos

Gahais replied:
yes, I moved team section to a whole new tab, and expanded these build line edits so that they are no longer cramped. I was planning that too to create team tab, after integrating Jack Matthews ideas for salvage in his GUI.
you are right about these disappearances of buttons, I pushed previous checkboxes a bit to the right before so there was overlapping of previous and next checkboxes. I found a good method to avoid such problems entirely
Can skip width and height of GUI elements and their size will be adjusted automatically so it removes problem of overlapping, i recommend that and I used this in below commits :)
`GUICtrlCreateCheckbox('Loop Runs', 31, 60)` instead of 
`GUICtrlCreateCheckbox('Loop Runs', 31, 60, 75, 25)`

Your idea of double axis table is good but I think below idea can be even better
I slept on it several nights and come up with this idea
Can upgrade this tree view to hold all the loot options not just loot components, so that there is one big expandable and collapsible tree view which can handle all the loot cases quite flexibly
For example it might be quite tough to put all the trophies into a 2D table, and hard to handle all the specific cases like store gemstones, but sell destroyer cores, and salvage glacial stones but store or sell dust from salvaging these stones
With loot options tree view it is elastic and can remove or add new branches any time without GUI modification each time, just with edition of loot conf json file, to update tree

So now when loot options tab is removed and new teams tab created there is the same number of 5 tabs so these tabs didn't take much space and could retain their full names, can reduce them to what you suggested in the future if there might be some new tab: main, run, loot, team, infos, (new tab here)
In the message below on the bottom there should be 4 screens of how this upgraded GUI looks like, i think you might like it :)

caustic-kronos commented on src/Farm-FoW.au3:
```diff
@@ -59,15 +59,44 @@ EndFunc
-	; Make party
-	; Assuming that team has been set up correctly manually
+	If GetMapID() <> $ID_Temple_of_the_Ages Then
+		If TravelToOutpost($ID_Temple_of_the_Ages, $DISTRICT_NAME) == $FAIL Then Return $FAIL
```
Combine in a single if

Gahais replied:
ok, I merged if setup statements in all farms in this commit 439463a5

caustic-kronos commented on src/Farm-FoWTowerOfCourage.au3:
```diff
@@ -0,0 +1,308 @@
+	If Not $FOW_TOC_FARM_SETUP Then
+		If SetupFoWToCFarm() == $FAIL Then Return $PAUSE
+	EndIf
+	If $STATUS <> 'RUNNING' Then Return $PAUSE
```
We can probably remove most of those. The idea was to allow user to pause after setup, but it's rarely used

caustic-kronos commented on src/Farm-FoWTowerOfCourage.au3:
```diff
@@ -0,0 +1,308 @@
+Func SetupFoWToCFarm()
+	Info('Setting up farm')
+	If GetMapID() <> $ID_Temple_of_the_Ages Then
+		If TravelToOutpost($ID_Temple_of_the_Ages, $DISTRICT_NAME) == $FAIL Then Return $FAIL
```
- [ ]  Let's merge the Ifs

caustic-kronos commented on src/Farm-GemstoneMargonite.au3:
```diff
@@ -0,0 +1,537 @@
+
+
+Func GemstoneMargoniteFarmLoop()
+	Sleep(2000)
```
Same comment concerning Sleep

Gahais replied:
ok, refactored this in commit 4fd122cc

caustic-kronos reviewed (changes requested):
Amazing !
Again, so many great things 😮 
Lots of good improvements in there so I only commented on issues/corrections/things I didn't understand.
Will need to give it several more look and lots of testing.
In the meantime here is already a first review with a lot of comments.
Good job !

Gahais commented on BotsHub.au3:
```diff
@@ -1028,7 +1028,7 @@ Func UpdateFarmDescription($Farm)
 			GUICtrlSetData($GUI_Label_FarmInformations, $SunspearArmorInformations)
 		Case 'Tasca'
-			GUICtrlSetData($GUI_Edit_CharacterBuild, $TascaChestRunnerSkillbar)
+			GUICtrlSetData($GUI_Edit_CharacterBuild, $TascaDervishChestRunnerSkillbar & '		' & $TascaAssassinChestRunnerSkillbar)
```
ok, I see that this might be a bit ugly
Probably a simple and good solution is to use List View that is available as inbuilt control in AutoIT, without much complicated fuss, like custom controls.
So can create a list view with fixed size in this farm informations tab, like to fit 1,5 or 2,5 template builds rows. This list view has an inbuilt scroll, so indefinite number of build templates can be appended to list view and scrolled in list view window
Below should be a screenshot how this list view looks
Another solution would be to use inbuilt Combobox, but it would show only 1 template build row and would require to click arrow near combobox to expand it
Another solution might be some custom controls but this can be complicated
<img width="1150" height="668" alt="screen_listview" src="https://github.com/user-attachments/assets/f7bd7a12-669c-412f-84d3-44e4f6b1c29c" />

Gahais replied:
ok, redesigned farm informations panel with multiedits in commit a7b33fbd, all content seems to fit quite nicely
below should be some screens
<img width="400" height="300" alt="Screenshot_20251227_170336" src="https://github.com/user-attachments/assets/778ada93-1d48-4e8d-983e-1c2eb01fe905" />
<img width="400" height="300" alt="Screenshot_20251227_170347" src="https://github.com/user-attachments/assets/3be914e1-2d4c-4612-80ea-05f93007fdbf" />
<img width="400" height="300" alt="Screenshot_20251227_171919" src="https://github.com/user-attachments/assets/cb97ba70-1fe9-4cfb-a68d-6d22e1981c1f" />

Gahais commented on src/Farm-GemstoneMargonite.au3:
```diff
@@ -0,0 +1,537 @@
+Func SetupPlayerMargoniteFarm()
+	Info('Setting up player build skill bar')
+	Sleep(500 + GetPing())
+	If DllStructGetData(GetMyAgent(), 'Primary') == $ID_Assassin Then
```
ok, good idea, will refactor that

Gahais replied:
All these if profession checks were refactored into switch cases in this commit https://github.com/caustic-kronos/BotsHub/commit/168d956e40744acc641de293b64b0521282ee02b

Gahais commented on src/Farm-Vaettirs.au3:
```diff
@@ -429,4 +490,110 @@ Func GetWastrelsTarget()
+	WEnd
+
+	Info('Killing Vaettirs')
+	If $VaettirsPlayerProfession == $ID_Assassin Or $VaettirsPlayerProfession == $ID_Mesmer Or $VaettirsPlayerProfession == $ID_Elementalist Then
```
ok, good idea, will refactor that in code places where this is used

Gahais replied:
All these if profession checks were refactored into switch cases in this commit https://github.com/caustic-kronos/BotsHub/commit/168d956e40744acc641de293b64b0521282ee02b

Gahais commented:
hi, oops, I didn't see these collapsed comments. I tried to expand these collapsed comments when logged out but nothing happened. It looks like that I need to be logged in to be able to expand collapsed comments. So I updated the project according to your comments :) I wrote in your comments how I refactored code according to your suggestions, you can read that when you have free time

This is my luck. They had to do Guild Wars Reforged update in the middle of my pull requests. Luckily I managed to farm 3 stacks of obsidians $$$ before bots stopped working :smiley: 
But it looks like that GUI can still be tested and modified even when bots aren't working currently, so I finished the upgrade of GUI. Now it is quite complete. Loot components are finally connected to the code. I also expanded this tree with all the loot options - pickup/salvage/sell/store, so that all the cases can be handled flexibly with a tree view. I think you might like it :)

I spoke with @JackLinesMatthews
Jack Matthews gave me permission to integrate his fork into master BotsHub project. So I integrated his Asura title farm and Minotaurs farm into BotsHub master. And I integrated his other ideas regarding pickup and salvage of items with some small modifications. I credited Jack Matthews accordingly in Credits.md and to files he contributed

- added checkbox and combobox for automatic switching of weapon slot for saved farm configurations
- added another radio button for buying faction scrolls for faction points in Kurzick/Luxon farms
- moved team options to a new tab
- removed loot options tab to move all these options to loot options tree view so that it contains all loot components + pickup/salvage/sell/store loot options and reworked utils-Storage code to handle loot according to GUI tree view settings. Added default loot options conf file
- finished these functions CreateValuableModsByWeaponTypeMap() and CreateValuableInscriptionsArray() so that bot should farm these components according with tree view settings
- refactored GUI so that is less cramped and has better layout of its elements, changed BotsHub version to 2.0
- added some function to traverse and check checkboxes up to the root if all children checkboxes became checked
- added some cast time effect calculation based on effect to improve functions like UseSkillEx
- added more IDs and dictionaries to Gwa2_id

You can review this pull request when you have free time, I sticked to your suggestions and programmed quite carefully + tested it so it should be rather fine :)
So you are planning to fix BotsHub after some other programmers come up with fixes to this GWA2 after Guild Wars Reforged update?
Let's not forget about this pull request, it has so much profit $$$ in it :smiley:

Gahais commented on src/Farm-GemstoneMargonite.au3:
```diff
@@ -0,0 +1,537 @@
+	Info('Casting hero monk bonds')
+	; below loops ensure that player have the effect of 5 monk enchantments from monk hero and also monk hero have 1 enchantment - balthazar's spirit
+	; last 2 enchantments are least important so these may deactivate when hero energy drops to 0, which is unlikely
+	While GetEffect($ID_Balthazars_Spirit) == Null
```
hmm, what do you think about this solution? Theoretical, mathematical calculation of how long exactly bot should be sleeping to recharge energy to cast these bond skills, usage of improved UseSkillEx() function. Tested it out and measured, looks to be fine. Added it to commit ab8cda82 with casting bonds


```
	UseSkillEx($Margonite_Hero_BalthazarSpirit, $PlayerAgentID, $MargoniteHeroIndex) ; costs 10 energy
	Sleep(10000) ; wait until energy is recovered, should recover 10 energy with 3 energy pips
	UseSkillEx($Margonite_Hero_WatchfulSpirit, $PlayerAgentID, $MargoniteHeroIndex) ; costs 15 energy
	UseSkillEx($Margonite_Hero_BlessedSignet, $MargoniteHeroAgentID, $MargoniteHeroIndex) ; recover 6 hero energy
	Sleep(12000) ; wait until Blessed signet is recharged, should recover 8 energy with 2 energy pips
	UseSkillEx($Margonite_Hero_LifeBarrier, $PlayerAgentID, $MargoniteHeroIndex) ; costs 15 energy, 1 energy should be recovered during casting, energy should be maxed
	UseSkillEx($Margonite_Hero_BlessedSignet, $MargoniteHeroAgentID, $MargoniteHeroIndex) ; recover 9 hero energy
	Sleep(15000) ; wait until Blessed signet is recharged, should recover 5 energy with 1 energy pip
	UseSkillEx($Margonite_Hero_LifeBond, $PlayerAgentID, $MargoniteHeroIndex) ; costs 10 energy
	UseSkillEx($Margonite_Hero_BlessedSignet, $MargoniteHeroAgentID, $MargoniteHeroIndex) ; recover 11 hero energy, energy should be maxed
	Sleep(10000) ; wait until Blessed signet is recharged, 0 pips
	UseSkillEx($Margonite_Hero_VitalBLessing, $PlayerAgentID, $MargoniteHeroIndex) ; costs 10 energy
	UseSkillEx($Margonite_Hero_BlessedSignet, $MargoniteHeroAgentID, $MargoniteHeroIndex) ; recover 11 hero energy
	Sleep(10000) ; wait until Blessed signet is recharged, around 3 energy lost with -1 pip
	UseSkillEx($Margonite_Hero_BalthazarSpirit, $MargoniteHeroAgentID, $MargoniteHeroIndex) ; costs 10 energy
	UseSkillEx($Margonite_Hero_BlessedSignet, $MargoniteHeroAgentID, $MargoniteHeroIndex) ; recover 11 hero energy, -2 pips, but energy will be recovered soon with balthazar's spirit

```

Gahais commented on src/Farm-Kournans.au3:
```diff
@@ -182,7 +198,7 @@ Func KournansFarmLoop()
 	;Sleep(1000)
 	UseSkillEx($Kournans_Earthquake, $targetFoe)	;3s
-	UseSkill($Kournans_DragonsStomp, $targetFoe)	;3s
+	UseSkillEx($Kournans_DragonsStomp, $targetFoe)	;3s
```
ok, sorry for this I dropped the commit that made this, so it is `UseSkill` as before

Gahais commented on lib/GWA2_ID.au3:
```diff
@@ -450,6 +450,23 @@ Global Const $ID_Mercenary_Hero_7	= 34
+										'Miku', 'ZeiRi']
+
+Global Const $HeroNamesFromIDs			=	MapFromArrays($HeroIDs, $HeroNames)
+Global Const $HeroIDsFromNames			=	MapFromArrays($HeroNames, $HeroIDs)
```
ok, it's good that you noticed that. You have some programmer's sixth sense :)
so I commented this line in commit 007aa5ae
`;Global Const $HeroNamesFromIDs`
Is this fine? We could keep these 2 dictionaries for completeness sake, but comment out the one which isn't used (I used the alternative `HeroIDsFromNames` for team options tab hero comboboxes)

Gahais commented on lib/GWA2_ID.au3:
```diff
@@ -232,6 +232,12 @@ Global Const $ID_Dervish		= 10
+Global Const $ID_Hero_fighting	= 0
+Global Const $ID_Hero_guarding	= 1
+Global Const $ID_Hero_avoiding	= 2
+#EndRegion Hero combat behaviour
```
hmm, `$ID_Hero_avoiding` is used to set hero behaviour in margonite gemstone farm so that monk doesn't aggro margonites. And in now added Glint challenge farm I used `$ID_Hero_fighting` for all heroes like pvx wiki article recommends it, so heroes aggro all destroyers

Gahais commented on lib/GWA2.au3:
```diff
@@ -3749,8 +3749,15 @@ EndFunc
+	Local $aftercast = DllStructGetData($skill, 'Aftercast') * 1000
+
+	; Caution, GetInstanceUpTime() may also include logging into account time which may not be included in in-game cast timestamps, therefore adding GetPing() + 3000 just in case and capping it to be always bigger or equal to 1 with _Max(), if Recharge is non-zero
+	Return $rechargeFutureTimestamp == 0 ? 0 : _Max($rechargeFutureTimestamp + $castTime + $aftercast + GetPing() + 3000 - GetInstanceUpTime(), 1)
```
well it was more like measurement, originally there wasn't much to pray for :) this GetSkillTimer probably worked better before some Guild Wars update
yeah, it works quite fine as GetInstanceUpTime() is quite similar to cast time, but difference could be suriprisingly negative sometimes. Maybe this function includes logging time but I don't guarantee that
ok, I removed this 3000 just in case now in the code to make better impression. If I come up with some better solutions I may improve these functions.
But it's quite fine. There is a safeguard that if player/hero's skill is already recharged then function returns 0, and second safeguard is capping recharge time remaining with minimal value 1 so that it won't get negative but stay at 1 which is often checked in the code that effect time is > 0 or >= 1 so it should be fine this way I think :)

Gahais commented on src/Farm-GemstoneStygian.au3:
```diff
@@ -98,32 +102,30 @@ Func SetupGemstoneStygianFarm()
-		LoadSkillTemplate($StygianAssasinSkillBar)
-	ElseIf DllStructGetData(GetMyAgent(), 'Primary') == $ID_Mesmer Then
+		LoadSkillTemplate($AMeStygianSkillBar)
+    ElseIf DllStructGetData(GetMyAgent(), 'Primary') == $ID_Mesmer Then
```
yeah, sorry, I come up with a trick that search project's most important files for 2 or 4 spaces and then checked and replaced them with tabs characters, so it should be cleaner in below commits

Gahais commented on lib/Utils.au3:
```diff
@@ -2508,98 +2508,98 @@ EndFunc
+	Local Static $PN_SS_Priest			= 6498
+	Local Static $PN_SS_Defender		= 6499
+	Local Static $PN_SS_Summoner		= 6507
+	Local Static $PN_Modniir_Priest		= 6512
```
yeah, I overlooked that. It is good that you noticed that, you have some programmer's sixth sense
So I dropped the commit that had this Static code, now it is without Static as before

Gahais commented on src/Farm-Lightbringer.au3:
```diff
@@ -225,51 +225,37 @@ Func SpeedTeam()
+	; Speed up team using Junundu Tunnel
 	; Get close enough to cast spells but not Aggro
-	; Use Junundu Siege (4) until it's in CD
+	; Use Junundu Siege (4) until it's in countdown
```
This is why I like long, descriptive names :laughing:
Changed this to cooldown in below commit

Gahais commented on lib/Utils.au3:
```diff
@@ -2199,30 +2199,91 @@ EndFunc
+
+
+;~ Detect if player is rubberbanding
+Func IsPlayerRubberBanding()
```
Dreams can come true :blush:

Gahais commented on src/Farm-Corsairs.au3:
```diff
@@ -71,7 +74,12 @@ Func CorsairsFarm($STATUS)
+	Resign()
+	Sleep(4000)
+	ReturnToOutpost()
+	Sleep(6000)
```
It is a very wise saying :) but I would make a sacrifice in this case for the sake of code coherence. I take responsibility, if it breaks, I can fix it
I improved this code in below commits, so that I created new smaller function ResignAndReturnToOutpost() which just resigns, waits these 3,5 seconds and returns, and all these farms use that function + ReturnBackToOutpost() in Utils also uses it so it is quite good code consistence and modularity
Now the code clearly points out that there are 6 farms, Gemstone Stygian/Margonite/Torment, Nexus/Sunspear challenges and Corsairs missions, which share the common situation that all of them have farm map id the same as outpost so can handle them with general function without the need of creating custom functions

Gahais commented on BotsHub.au3:
```diff
@@ -1146,6 +1146,10 @@ EndFunc
+	; below line is a fix for a very weird bug that character combobox truly updates only after being set second time. _JSON_Get() function seems to be fine, maybe this is AutoIT bug
+	GUICtrlSetData($GUI_Combo_CharacterChoice, _JSON_Get($jsonObject, 'main.character'))
+	GUICtrlSetData($GUI_Combo_FarmChoice, _JSON_Get($jsonObject, 'main.farm'))
+	; below line is a fix for a very weird bug that farm combobox sometimes updates only after being set second time. _JSON_Get() function seems to be fine, maybe this is AutoIT bug
```
yeah, this was some weird glitch, probably an AutoIT bug of GUI. The GUI can still be tested even if bots aren't working so you can test it out if you want :)

caustic-kronos replied:
I think I might have found your issue (not entirely sure). In any case, your fix somehow breaks one behaviour I need from bot.
Ok, so basically, I made it so that when you start the bot, if GW is started, then it picks up character's name instantly (that saves me 2 clicks everytime I start the bot, which is a lot when testing code). This is done in:
`	GUICtrlSetData($GUI_Combo_CharacterChoice, $comboList, $gameClients[0][0] > 0 ? $gameClients[1][3] : '')`
Basically, if there is at least a client started, we use that client currently logged character.
Problem for you I believe is that it overrides the character set inside configuration and somehow setting it twice correct that ? 😱 Not quite sure there, but it should be something along those lines.

Gahais commented on BotsHub.au3:
```diff
@@ -169,155 +180,218 @@ Global $GUI_Label_ToDoList
+	$GUI_Checkbox_BuyEctoplasm = GUICtrlCreateCheckbox('Buy ectoplasm', 31, 315, 85, 20)
+	$GUI_Checkbox_BuyObsidian = GUICtrlCreateCheckbox('Buy obsidian', 205, 315, 75, 20)
+	$GUI_Checkbox_FarmMaterials = GUICtrlCreateCheckbox('Farm materials', 31, 340, 85, 20)
+	$GUI_Checkbox_CollectData = GUICtrlCreateCheckbox('Collect data', 205, 340, 75, 20)
```
yeah, sorry for the break. I was thinking that, after these 2 pull request with BotsHub version 2.0, lot of stuff was changed in upgraded GUI, so would need to create new configuration files again nevertheless
It's good that you noticed about this mismatched order of item management actions. ok, I refactored that. New GUI solved everything. I think you may like it :)
I added some warning description to Readme
"Users are advised to create and save new configuration files using GUI.  
Old configuration files may not be backward compatible" 
Users have been warned
:smile:

Gahais replied:
It's good that you noticed about this mismatched order of item management actions. I refactored that. New GUI solved everything. I think you may like it :)
Below at the bottom message there should be 4 screens of main tabs of GUI

Gahais commented on src/Farm-GemstoneMargonite.au3:
```diff
@@ -0,0 +1,537 @@
+Func GemstoneMargoniteFarm($STATUS)
+	; Need to be done here in case bot comes back from inventory management
+	If Not $GEMSTONE_MARGONITE_FARM_SETUP Then
+		If SetupGemstoneMargoniteFarm() == $FAIL Then Return $PAUSE
```
ok, I merged if setup statements in all farms in this commit 439463a5

Gahais commented on src/Farm-Lightbringer2.au3:
```diff
@@ -51,7 +51,7 @@ EndFunc
 Func Lightbringer2FarmSetup()
 	Info('Setting up farm')
-	TravelToOutpost($ID_Kodash_Bazaar, $DISTRICT_NAME)
+	If GetMapID() <> $ID_Kodash_Bazaar Then TravelToOutpost($ID_Kodash_Bazaar, $DISTRICT_NAME)
```
ok, I refactored in commit ca904118 all these travel calls so that almost nowhere these map checks are used and code is much more concise now 
There is only one check in `TravelToOutpost()` if player is already in the location and then that function immediately returns `$SUCCESS` without printing any unnecessary messages, so this `TravelToOutpost()` can be called anywhere frivolously without impact to make sure player is in the desired location
And `TravelToOutpost()` now waits `$maxWaitTime=10000` interval to try to get to location and returns `$FAIL` if max time elapsed

Gahais commented on src/Farm-FoWTowerOfCourage.au3:
```diff
@@ -0,0 +1,308 @@
+Global $FowToCFarmTimer = Null
+Global $FOW_TOC_FARM_SETUP = False
+
+Global $FoWToCMoveOptions = CloneDictMap($Default_MoveDefend_Options)
```
hmm, do you mean `Global Const $FoWToCMoveOptions = CloneDictMap($Default_MoveDefend_Options)` ?
hmm, but below there are calls like this one, something more or less like edition 
`$FoWToCMoveOptions.Item('defendFunction') = DefendFoWToC`
Maybe we could leave this without Const

Gahais commented on src/Farm-FoWTowerOfCourage.au3:
```diff
@@ -0,0 +1,308 @@
+
+
+Func CastBuffsFowToC()
+	If IsPlayerDead() Then Return $FAIL
```
ok, you are right that I might have used IsPlayerAlive() checks quite often
I removed a lot of redundant IsPlayerAlive() checks in this commit 877b52a4
But for this particular function `CastBuffsFowToC()` I would leave this check at the top
I removed this check from below function `DefendFoWToC()`
Because I think it would be good idea to check to ensure player status in `CastBuffsFowToC()` function before doing calls like IsRecharged() or `UseSkillEx()`. I think it might be better to check than to do such calls on potentially invalid sate when player is dead. I suspect that such actions on invalid states might lead to crashes
I removed a lot of these checks in other places

Gahais commented on src/Farm-GemstoneMargonite.au3:
```diff
@@ -0,0 +1,537 @@
+
+
+Func DisableMargoniteHeroSkills()
+	Sleep(500)
```
ok, you are right, it was quite weird. So I removed a lot of redundant sleep calls or moved them into better places in this commit 4fd122cc

Gahais commented on src/Farm-GemstoneMargonite.au3:
```diff
@@ -0,0 +1,537 @@
+	Sleep(2000)
+	$MargoniteHeroAgentID = GetHeroID($MargoniteHeroIndex)
+	If Not GetAgentExists($MargoniteHeroAgentID) Then Return $FAIL
+	If IsPlayerDead() Then Return $FAIL
```
ok, I removed a lot of IsPlayerAlive/Dead checks in commit 877b52a4
I used this check to ensure that map and everything loaded properly but now I wrote better check
`If Not GetAgentExists($PlayerAgentID) Or Not GetAgentExists($MargoniteHeroAgentID) Then Return $FAIL`

Gahais commented on src/Farm-GemstoneMargonite.au3:
```diff
@@ -0,0 +1,537 @@
+		Sleep(1000)
+		Dialog(0x82EF01)
+		Sleep(1000)
+	Until GetQuestByID(0x2EF) <> Null Or TimerDiff($TimerQuest) > 15000 Or IsPlayerDead()
```
ok, I like sticking to While loop convention too regardless of 1 redundant check
I changed that to While loop in this commit 3a948c1b

Gahais commented on src/Farm-GemstoneMargonite.au3:
```diff
@@ -0,0 +1,537 @@
+		If IsRecharged($Margonite_ShroudOfDistress) And Not IsRecharged($Margonite_ShadowForm) And GetEnergy() > 14 Then UseSkillEx($Margonite_ShroudOfDistress)
+	EndIf
+
+	If $MargonitePlayerProfession == $ID_Assassin Then
```
All these if profession checks were refactored into switch cases in this commit https://github.com/caustic-kronos/BotsHub/commit/168d956e40744acc641de293b64b0521282ee02b

Gahais commented on src/Farm-Vaettirs.au3:
```diff
@@ -78,7 +78,7 @@ Func RunToJagaMoraine()
 
 	Info('Running to Jaga Moraine')
-	Local $pathToJaga[30][2] = [ _
+	Local Static $pathToJaga[30][2] = [ _
```
ok, I dropped the commit which had these Static keywords so semantics dilemma should be resolved :)

Gahais commented on src/Farm-Corsairs.au3:
```diff
@@ -82,27 +88,41 @@ Func SetupCorsairsFarm()
-	SetupTeamCorsairsFarm()
-	LoadSkillTemplate($RACorsairsFarmerSkillbar)
+	If SetupPlayerCorsairsFarm() == $FAIL Then Return $FAIL
+	If SetupTeamCorsairsFarm() == $FAIL Then Return $FAIL
```
All this stuff is now in the commit 9d7898be

Gahais commented on BotsHub.au3:
```diff
@@ -1138,6 +1138,7 @@ Func WriteConfigToJson()
 	_JSON_addChangeDelete($jsonObject, 'loot.consumables.trick_or_treat_bags', GUICtrlRead($GUI_Checkbox_LootToTBags) == 1)
 	_JSON_addChangeDelete($jsonObject, 'loot.consumables.candy_cane_shards', GUICtrlRead($GUI_Checkbox_LootCandyCaneShards) == 1)
 	_JSON_addChangeDelete($jsonObject, 'loot.consumables.lunar_tokens', GUICtrlRead($GUI_Checkbox_LootLunarTokens) == 1)
+	_JSON_addChangeDelete($jsonObject, 'run.bags_count', Number(GUICtrlRead($GUI_Combo_BagsCount)))
```
yeah, sorry for the break. I was thinking that, after these 2 pull request with BotsHub version 2.0, lot of stuff was changed in upgraded GUI, so would need to create new configuration files again nevertheless
I added some warning description to Readme
"Users are advised to create and save new configuration files using GUI.
Old configuration files may not be backward compatible"
Users have been warned
:smile:

Gahais commented on src/Farm-GlintChallenge.au3:
```diff
@@ -0,0 +1,212 @@
+
+	If IsPlayerAndPartyWiped() Or GetIsDead(GetAgentById($AgentID_BabyDragon) Then Return $FAIL
+
+	; Tripled to try securing the looting of Chest of the Bortherhood
```
refactored it

Gahais commented on BotsHub.au3:
```diff
@@ -1231,7 +1231,7 @@ Func UpdateFarmDescription($Farm)
 			GUICtrlSetData($GUI_Label_FarmInformations, $SunspearArmorInformations)
 		Case 'Tasca'
-			GUICtrlSetData($GUI_Edit_CharacterBuild, $TascaDervishChestRunnerSkillbar & '		' & $TascaAssassinChestRunnerSkillbar)
+			GUICtrlSetData($GUI_Edit_CharacterBuild, $TascaAssassinChestRunnerSkillbar)
```
Dervish comes back in commit a7b33fbd :) along with these multi edit farm descriptions

Gahais commented on src/Farm-DragonMoss.au3:
```diff
@@ -170,8 +170,10 @@ Func DragonMossFarmLoop()
 
-	Info('Looting')
-	PickUpItems()
+	If IsPlayerAlive() Then
```
ok, refactored this in commit f2e09a0a

Gahais commented on lib/GWA2.au3:
```diff
@@ -2233,6 +2233,45 @@ Func ChangeWeaponSet($weaponSet)
+	For $effect in $effects
+		$effectID = DllStructGetData($effect, 'EffectId')
+		; consumables effects
+		If $effectID == $ID_Essence_of_Celerity_effect Then $castTime = 0.80 * $castTime
```
ok, refactored this in commit f2631a85

Gahais commented on lib/GWA2.au3:
```diff
@@ -1663,10 +1663,12 @@ EndFunc
-	If $target <> 0 Then $targetId = DllStructGetData($target, 'ID')
-	DllStructSetData($useHeroSkillStruct, 2, GetHeroID($hero))
+	If $target <> 0 And IsDllStruct($target) Then $targetId = DllStructGetData($target, 'ID') ; case when target agent struct was passed to function
+	If $target <> 0 And Not IsDllStruct($target) Then $targetId = $target ; case when target ID was passed to function
```
ok, refactored this into acceptance only of structs and Nulls in commit c4596a8b

Gahais commented on lib/GWA2.au3:
```diff
@@ -2273,26 +2275,51 @@ EndFunc
+		Sleep(50 + GetPing())
+	Until (Not IsRecharged($skillSlot, $heroIndex)) Or ($fullCastTime < TimerDiff($castTimer))
+	Return True
 EndFunc
```
ok, refactored this in commit fa593aa3

Gahais commented on src/Farm-GemstoneMargonite.au3:
```diff
@@ -40,8 +40,9 @@ Global Const $MargoniteMonkHeroSkillBar = 'OwITAnHb5Qe/zhxLkpE6+G'
-Global $MargonitePlayerProfession = $ID_Mesmer ; global variable to remember player's profession in setup and avoid creating Dll structs over and over during fight
+Global $PlayerAgentID = Null ; agent ID that is randomly assigned to player in explorable areas
+Global $MargoniteHeroAgentID = Null ; agent ID that is randomly assigned to hero in explorable areas
+Global $MargonitePlayerProfession = $ID_Mesmer ; global variable to remember player's profession in setup to avoid creating Dll structs over and over
```
hey, it turns out that these 2 lines got removed from the code after all :) in commit bf11943f
`Global $PlayerAgentID = Null ; agent ID that is randomly assigned to player in explorable areas`
`Global $MargoniteHeroAgentID = Null ; agent ID that is randomly assigned to hero in explorable areas`
because of cleaner code elsewhere that now IsHeroDead and IsHeroAlive and UseSkillTimed which gets structs, are used instead so there is no need to remember hero agent ID or playerID

Gahais commented on src/Farm-Boreal.au3:
```diff
@@ -124,8 +124,14 @@ Func SetupPlayerBorealChestFarm()
-	;ChangeWeaponSet(1) ; change to other weapon slot or comment this line if necessary
-	Sleep(500 + GetPing())
+	Sleep(250 + GetPing())
+	If GUICtrlRead($GUI_Checkbox_WeaponSlot) == $GUI_CHECKED Then
```
ok, moved out setting up player, team and weapon slot out of farms in generalized functions in Utils in commits c5c86821 and 5b4c78bc

caustic-kronos commented on conf/loot/Default Loot Configuration.json:
```diff
@@ -2,56 +2,364 @@
-				"Sword": false,
-				"Wand": false
+				"Axe": {
+					"Req 8": false,
```
Just for information: q8 only loot max stats for focus, shields and swords, not for anything else.
So axes, bows, daggers, etc... will loot in q8 but without max damage.
If we add all req then it doesn't matter anyway.

Gahais replied:
ok, thanks for the information :)
so leave it as it is? without changing?

Gahais replied:
ok, all weapon Req from 0 to 13 are added for all items in commit b20859c6 , except green items, without Reqs
I added an exception for purple, gold items with Req 8 for focus, shields and swords as you mentioned
So that these in exception are picked up, not sold and stored :)

caustic-kronos commented on lib/Utils-Storage-Bot.au3:
```diff
@@ -781,7 +801,7 @@ Func IsUnidentifiedGoldItem($item)
 
 
-;~ helper function for StoreEverythingInXunlaiStorage function
+;~ helper function for StoreEverythingInXunlaiStorage() function
```
I think it's preferable to not put parentheses when referencing a function like that.
When you want to look for a function call on an simple editor you can search for "function("

caustic-kronos commented on BotsHub.au3:
```diff
@@ -1543,6 +1565,48 @@ Func IterateOverTreeView(ByRef $context, $treeViewHandle, $treeViewItem = Null,
+	Local $treeViewItem
+	; checking if $startingPoint contains extended path separated with . characters or is single tree element
+	If StringInStr($startingPoint, '.') > 0 Then ; $startingPoint contains extended path, not single tree element
+		Opt('GUIDataSeparatorChar', '.')
```
And as I thought, having both is basically super dangerous. Well, at least if you are not very careful.
This:
	```Opt('GUIDataSeparatorChar', '|') ;'|' is the default, but caution is advised that it can be overriden by '.' which is used for tree view```
is run everytime we change the character in the character combo. That created a funny bug, where if I selected a character via the combobox, it broke looting because suddenly we used '|' instead of '.'.
If character is "by default" (either because we have single instance running or because it's the default character from configuration), then loot happens just fine.

In any, let's try avoiding those funny bugs by not doing switching from '.' to '|' and vice versa. We should stick to one, and if possible keep it the original '|', since it's the default. There must be a way to use '.' without changing GUIDataSeparatorChar.

Gahais replied:
ok, maybe something like this?

```
Func FindNodeInTreeView($treeViewHandle, $treeViewItem = Null, $currentPath = '', $pathDelimiter = '.')
	If $treeViewItem == Null Then $treeViewItem = _GUICtrlTreeView_GetFirstItem($treeViewHandle)

	Local $pathArray = StringSplit($currentPath, $pathDelimiter)
	; deleting first element of path array (which has the count of characters in AutoIT) to have array indexed from 0
	_ArrayDelete($pathArray, 0)
	If UBound($pathArray) == 0 Then Return Null

	$currentPath = StringTrimLeft($currentPath, StringLen($pathArray[0]) + 1)

	While $treeViewItem <> 0
		If UBound($pathArray) == 1 And $pathArray[0] == _GUICtrlTreeView_GetText($treeViewHandle, $treeViewItem) Then Return $treeViewItem
		If UBound($pathArray) > 1 And $pathArray[0] == _GUICtrlTreeView_GetText($treeViewHandle, $treeViewItem) And _
			_GUICtrlTreeView_GetChildCount($treeViewHandle, $treeViewItem) > 0 Then Return FindNodeInTreeView($treeViewHandle, _GUICtrlTreeView_GetFirstChild($treeViewHandle, $treeViewItem), $currentPath)
		$treeViewItem = _GUICtrlTreeView_GetNextSibling($treeViewHandle, $treeViewItem)
	WEnd
	Return Null
EndFunc
```

Gahais replied:
ok, I did a thorough refactor of this tree traversal stuff in commit cc53de54, created a custom tree view item finding function `FindNodeInTreeView()` like above
I also refactored `IterateOverTreeView()` function to be a bit more compact, in similar way as `FindNodeInTreeView()`
tested these functions on some example tree and seem to be fine
Any path delimiter character can be used for loot tree view traversal, by default it is dot `.`
So there is no collision between loot tree view path delimiter and comboboxes data separators `|`
Only 1 default data separator is used for entire project `|` for comboboxes in GUI :)

Gahais commented on lib/Utils.au3:
```diff
@@ -1061,7 +1061,7 @@ EndFunc
 	Local $arraySize = UBound($itemArray)
 	Local $counts[$arraySize]
-	For $bagIndex = 1 To $BAGS_COUNT
+	For $bagIndex = 1 To 5
```
ok, I can revert it with rollback :)
but are you sure? if user puts war supplies or gemstones into 4th or 5th bag then they won't be counted in GUI, and counting shouldn't be harmful

Gahais commented on src/Farm-Boreal.au3:
```diff
@@ -125,13 +125,7 @@ Func SetupPlayerBorealChestFarm()
-		Info('Automatic player weapon slot setting is disabled. Assuming that player sets weapon slot manually')
-	EndIf
-	Sleep(250 + GetPing())
+	TrySetupWeaponSlotUsingGUISettings()
```
hmm, you mean something like that in BotsHub.au3?
```
		Case 'Boreal'
			$INVENTORY_SPACE_NEEDED = 5
			TrySetupPlayerUsingGUISettings()
			TrySetupTeamUsingGUISettings()
			$GLOBAL_FARM_SETUP = True
			$result = BorealChestFarm($STATUS)
```

I can try it but are you sure about it?
In some farms like Vaettirs or Spirit Slaves bot starts from some run place and shouldn't try to setup teams or player builds
I think each farm can be quite specific and each of them should be handled in unique way
I think not doing farm setup in BotsHub.au3 can be quite a good separation of concerns, to let farm files concern with farm setup

caustic-kronos replied:
Yeah, let me be clearer.
Basically, I suggest having two level setup :
- one general setup, valid for all bots, including all the things that can be valid for all bots
- one farm specific setup, the way we currently have

So that general setup can include weapon swap: I see no problem in including that even in farms with specific weapons/weapon swaps.
I think it could also include heroes and hero skills. For farms requiring a specific hero team, that team will be setup AFTER the "general" setup anyway, which means that character will leave group and add the hero(es) needed. But for farms requiring a full team and that don't automatically add it (because it's user specific, like all dungeon or vanquish bots), then having the team setup done at this point will be useful.

Here's how it currently work:
1- start global bot
2- global bot checks inventory for inventory management
3- global bot calls bot specific bot
4- specific bot checks if it's set up, if it's not, it runs the specific setup

I just suggest to add between step 1 and 2 a new step: a general setup that will setup weapon swap and heroes team.

Gahais commented on lib/Utils.au3:
```diff
@@ -306,56 +306,79 @@ EndFunc
+		Local $dyeColorID = DllStructGetData($item, 'DyeColor')
+		Local $dyeColorName = $DyeNamesFromIDs[$dyeColorID]
+		Return IsLootOptionChecked('Pick up items.Dyes.' & $dyeColorName)
+	ElseIf IsTrophy($itemID) Then
```
ok, this is refactored in commit ebc852c4

Gahais commented on lib/Utils-Storage-Bot.au3:
```diff
@@ -102,12 +102,12 @@ Func InventoryManagementBeforeRun()
 	; 10-Store items
 	If GUICtrlRead($GUI_Checkbox_StoreUnidentifiedGoldItems) == $GUI_CHECKED Then
-		TravelToOutpost($ID_Eye_of_the_North, $DISTRICT_NAME)
+		TravelToOutpost($ID_Embark_Beach, $DISTRICT_NAME)
```
ok, I created a utility function NPCCoordinatesInTown()
It receives town ID and NPC type and it returns coordinates of that NPC in that town
Any NPC can be used, Dye trader, Scroll trader, etc. The function can be expanded
So for trading Embark Beach or Eye of the North can be used, coordinates are added to function

Embark Beach has an advantage that it never gets decorated, while Eye of the North gets decorated in Christmas, so it swaps map ID. Need to do some calendar time checks, and it can be problematic to convert BotsHub timezone to Guild Wars server timezone for start and end hours of festivals
Also it looks like each character has to visit decorated town first before being able to travel there automatically with bots, even when visited before festival. Each decorated map seems to be considered as unknown outpost until visited. Embark Beach can be accessible early in campaigns as far as I know
I would vote for Embark Beach, it's a hot place :sunglasses: :palm_tree: 
Or can do a specific workaround that can travel to buy/sell items to Embark Beach only in second half of December and first half of January and rest of the year to Eye of the North

Gahais commented:
ok, thanks for the merge and fixes :)
